src/lib/Imp.hs

```diff
@@ -1538,25 +1535,61 @@ liftBuilderImp cont = do
   dropSubst $ translateDeclNest decls $ substM result
 {-# INLINE liftBuilderImp #-}
 
--- TODO: should we merge this with `liftBuilderImp`? Not much harm in
--- simplifying even if it's not needed.
-liftBuilderImpSimplify
-  :: Emits n
-  => (forall l. (Emits l, DExt n l) => BuilderM l (Atom l))
-  -> SubstImpM i n (Atom n)
-liftBuilderImpSimplify cont = do
-  block <- dceApproxBlock <$> buildBlockSimplified cont
-  dropSubst $ translateBlock Nothing block
-{-# INLINE liftBuilderImpSimplify #-}
+-- === Type classes ===
 
 unsafeFromOrdinalImp :: Emits n => IxType n -> IExpr n -> SubstImpM i n (Atom n)
-unsafeFromOrdinalImp ixTy i = liftBuilderImpSimplify do
+unsafeFromOrdinalImp (IxType _ dict) i = do
   i' <- toScalarAtom =<< sinkM i
-  unsafeFromOrdinal (sink ixTy) i'
+  liftEnvReaderM (projPreSimpDictThunk dict 2) >>=
+    (`appReduceImp` (Con $ Newtype NatTy i'))
 
 indexSetSizeImp :: Emits n => IxType n -> SubstImpM i n (IExpr n)
-indexSetSizeImp ty = fromScalarAtom =<< liftBuilderImpSimplify do
-  indexSetSize $ sink ty
+indexSetSizeImp (IxType _ dict) = fromScalarAtom . unwrapNewtype =<< case dict of
+  DictCon (IxFin n) -> return n
+  _ -> liftEnvReaderM (projPreSimpDictThunk dict 0) >>= (`appReduceImp` UnitVal)
+
+indexSetSizePreSimp :: Emits n => IxType n -> BuilderM n (Atom n)
+indexSetSizePreSimp (IxType _ dict) = unwrapNewtype <$> case dict of
+  DictCon (IxFin n) -> return n
+  _ -> liftEnvReaderM (projPreSimpDictThunk dict 0) >>= (`appReduce` UnitVal)
+
+ordinalPreSimp :: Emits n => IxType n -> Atom n -> BuilderM n (Atom n)
+ordinalPreSimp (IxType _ dict) i =
+  liftM unwrapNewtype $ liftEnvReaderM (projPreSimpDictThunk dict 1) >>= (`appReduce` i)
+
+unsafeFromOrdinalPreSimp :: Emits n => IxType n -> Atom n -> BuilderM n (Atom n)
+unsafeFromOrdinalPreSimp (IxType _ dict) o =
+  liftEnvReaderM (projPreSimpDictThunk dict 2) >>= (`appReduce` (Con $ Newtype NatTy o))
+
+projPreSimpDictThunk :: Atom n -> Int -> EnvReaderM n (Atom n)
```

Comment:
I found this "preSimp" terminology confusing. I think it's meant to be read as "presimplified" or "previously simplified", but I read it as "pre simplification" as in "this works with IxTypes that haven't been simplified yet. Maybe change to "projDictThunkFromSimp" or something?